File: src/Valid Parentheses/Valid Parentheses.py
class Solution:
	# @return a boolean
	def isValid(self, s):
		dic = {')':'(', ']':'[', '}':'{'}
		stack = []

		for c in s:
			if c in dic:
				if len(stack) > 0 and stack[-1] == dic[c]:
					# stack not empty & last element has matched
					stack.pop()
				else:
					return False
			else:
				stack.append(c)
		# check if stack is empty or not, return True if stack is empty
		return not stack

File: src/Valid Parentheses/Valid Parentheses.cpp
class Solution {
	public:
		bool isValid(string s) 
		{
			map<char, char> pairing;
			pairing[')'] = '(';
			pairing[']'] = '[';
			pairing['}'] = '{';

			stack<char> mystack;
			for(int i = 0 ; i < s.length() ; i++)
			{
				char c = s[i];
				if(c == '(' || c == '[' || c == '{')
				{
					mystack.push(c);
				}
				else
				{
					if( !mystack.empty() && ( mystack.top() == pairing[c] ) )
					{
						mystack.pop();
					}
					else
					{
						return false;
					}
				}
			}

			if(mystack.empty()) return true; //stack empty
			else return false;

		}
};
